src/VscProcess.cpp

```diff
@@ -403,75 +413,81 @@ void VscProcess::readFromVehicle()
-        break;
-      case MSG_USER_FEEDBACK:
-        //			handleFeedbackMsg(&recvMsg);
-        break;
-      case MSG_SETUP_KEY_INT_2:
-        //			handleGetSettingInt2(&recvMsg);
-        break;
-      case MSG_SETUP_KEY_INT_1:
-        if(handleGetSettingInt(recvMsg) == 0)
-        {
-          lastDataRx = ros::Time::now();
-        }
-        break;
-      case MSG_SETUP_KEY_STRING_1:
-        if(handleGetSettingString(recvMsg) == 0)
-        {
-          lastDataRx = ros::Time::now();
-        }
-        break;
-      default:
-        ROS_WARN("Received Invalid Message of type: %02x", recvMsg.msg.msgType);
-        errorCounts.invalidRxMsgCount++;
-        break;
+      /* Read next Vsc Message */
+      switch (recvMsg.msg.msgType)
+      {
+        case MSG_VSC_HEARTBEAT:
+          if (handleHeartbeatMsg(recvMsg) == 0)
+          {
+            lastDataRx = ros::Time::now();
+          }
+          break;
+        case MSG_VSC_JOYSTICK:
+          if (joystickHandler->handleNewMsg(recvMsg) == 0)
+          {
+            lastDataRx = ros::Time::now();
+          }
+          break;
+        case MSG_VSC_REMOTE_STATUS:
+          if(handleRemoteStatusMsg(recvMsg) == 0)
+          {
+            lastDataRx = ros::Time::now();
+          }
+          break; 
+        case MSG_VSC_NMEA_STRING:
+          //			handleGpsMsg(&recvMsg);
+          break;
+        case MSG_USER_FEEDBACK:
+          //			handleFeedbackMsg(&recvMsg);
+          break;
+        case MSG_SETUP_KEY_INT_2:
+          //			handleGetSettingInt2(&recvMsg);
+          break;
+        case MSG_SETUP_KEY_INT_1:
+          if(handleGetSettingInt(recvMsg) == 0)
+          {
+            lastDataRx = ros::Time::now();
+          }
+          break;
+        case MSG_SETUP_KEY_STRING_1:
+          if(handleGetSettingString(recvMsg) == 0)
+          {
+            lastDataRx = ros::Time::now();
+          }
+          break;
+        default:
+          ROS_WARN("Received Invalid Message of type: %02x", recvMsg.msg.msgType);
+          errorCounts.invalidRxMsgCount++;
+          break;
+      }
     }
   }
 
   // Log warning when no data is received
   ros::Duration noDataDuration = ros::Time::now() - lastDataRx;
-  if (noDataDuration > ros::Duration(.25))
+  ros::Duration reconnectTimeout = ros::Time::now() - lastReconnectAttempt;
+  if (noDataDuration > ros::Duration(.25) && reconnectTimeout > ros::Duration(5.0))
   {
     ROS_WARN_THROTTLE(.5, "No Data Received in %i.%09i seconds", noDataDuration.sec, noDataDuration.nsec);
     ROS_WARN_STREAM("Attempting to reconnect to the VSC...");
+    
+    lastReconnectAttempt = ros::Time::now();
+
     /* Open VSC Interface */
     vscInterface = vsc_initialize(serial_port_.c_str(), serial_speed_);
     if (vscInterface == NULL)
     {
-      ROS_FATAL("Cannot open serial port! (%s, %i)", serial_port_.c_str(), serial_speed_);
+      ROS_ERROR_THROTTLE(0.5,"Cannot open serial port! (%s, %i)", serial_port_.c_str(), serial_speed_);
```

Comment:
I realize this was here before your changes, but it feels like it should attempt to connect before reading. I guess it is fine as is.

Reply:
I decided to make as minimal changes as possible to the existing codebase, and they already had a part here where they checked for connectivity. I thought it would be best to insert the reconnect functionality here, where the segfault was originating.